+nosnoc/+time_freezing/cls_inelastic_multicontact.m

```diff
@@ -0,0 +1,293 @@
+                    beta_prod_expr_guess(ii-1) = beta_prod_expr_guess(ii-2)*(alpha_qv(ii)); % this is to have an expression depending only on alpha for the inital guess eval
+                end
+                alpha_ode = beta_prod(end)*(alpha_qv(dims.n_contacts)); % last lifting term;
+            end
+            % lifting of aux dyn multiplier expressions
+            for ii = 1:dims.n_contacts
+                if cls_model.friction_exists
+                    alpha_aux(ii) = (1-alpha_qv(ii))*(alpha_v_tangent(ii));
+                    alpha_aux(dims.n_contacts+ii) = (1-alpha_qv(ii))*(1-alpha_v_tangent(ii));
+                else
+                    alpha_aux(ii)=(1-alpha_qv(ii));
+                end
+            end
+        else
+            % with two smooth switching functions
+            beta_bilinear_ode_expr = alpha_q.*alpha_v_normal;
+            if cls_model.friction_exists
+                beta_bilinear_aux_expr = (1-alpha_q).*(1-alpha_v_normal);
+            end
+
+            if dims.n_contacts <= 2
+                % here no lifting of product terms
+                for ii = 1:dims.n_contacts
+                    alpha_ode = alpha_ode*(alpha_q(ii)+alpha_v_normal(ii)-beta_bilinear_ode(ii));
+                end
+            else
+                % here lifting of product terms
+                g_z_tf_beta_prod  = [beta_prod(1) - (alpha_q(1)+alpha_v_normal(1)-beta_bilinear_ode(1))*(alpha_q(2)+alpha_v_normal(2)-beta_bilinear_ode(2))]; % first lifting terms
+                                                                                                                                                              % lifting terms in between
+                for ii = 3:dims.n_contacts-1
+                    beta_prod_expr(ii-1) = beta_prod(ii-2)*(alpha_q(ii)+alpha_v_normal(ii)-beta_bilinear_ode(ii));
+                    beta_prod_expr_guess(ii-1) = beta_prod_expr_guess(ii-2)*(alpha_q(ii)+alpha_v_normal(ii)-beta_bilinear_ode(ii));
+                end
+                % last lifting term;
+                alpha_ode = beta_prod(end)*(alpha_q(dims.n_contacts)+alpha_v_normal(dims.n_contacts)-beta_bilinear_ode(dims.n_contacts));
+            end
+            % lifting of aux dyn multiplier expressions
+            for ii = 1:dims.n_contacts
+                if cls_model.friction_exists
+                    alpha_aux(ii) = beta_bilinear_aux(ii)*(alpha_v_tangent(ii));
+                    alpha_aux(dims.n_contacts+ii) = beta_bilinear_aux(ii)*(1-alpha_v_tangent(ii));
+                else
+                    alpha_aux(ii) = (1-alpha_q(ii))*(1-alpha_v_normal(ii));
+                end
+            end
+        end
+    end
+    heaviside_model.z = [theta;beta];
+    heaviside_model.g_z = [theta-[alpha_ode;alpha_aux];
+        beta - [beta_bilinear_ode_expr; beta_bilinear_aux_expr; beta_prod_expr]];
+    
+    f_all = [f_ode (inv_M_ext*f_aux)];
```

Comment:
here the inv_M_ext is used instead of inv_M on every auxiliary. I like the readiblity of the latter more, and it should not have any influence on the perfomance.